### ISSUE TITLE
Add SOLD2 line detector/descriptor/matcher to kornia

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#1124]
## 🚀 Feature
It is sota and implemented in pytorch under MIT license
https://github.com/cvg/SOLD2

**Comment by ducha-aiki** [comment]
cc author: @rpautrat

**Comment by rpautrat** [comment]
Great idea! I can start working on it starting in August, let's discuss it at that time.

**Comment by ducha-aiki** [comment]
Done in https://github.com/kornia/kornia/pull/1844

**Comment by ogencoglu** [comment]
@rpautrat  Is there a way to get the line segment probability outputs for each segment?

**Comment by rpautrat** [comment]
Hi, yes you could obtain a line score by interpolating the line heatmap along each segment. In pytorch, this could look like this:
```
def get_line_score(line_segments, line_heatmap, n_samples=5):
    h, w = line_heatmap.shape[1:3]
    samples = line_segments[:, :, 0, None] + torch.linspace(n_samples, device=line_segments.device).reshape(1, 1, n_samples) * (line_segments[:, :, 1, None] - line_segments[:, :, 0, None])  # samples is (B, N_LINES, N_SAMPLES, 2)
    samples = 2 * samples[:, :, :, [1, 0]] / torch.tensor([w, h]).to(samples) - 1
    sampled_scores = torch.nn.functional.grid_sample(line_heatmap[:, None], samples)[:, 0]
    return sampled_scores.mean(dim=2)  # shape (B, N_LINES)
```

**Comment by edgarriba** [comment]
@ogencoglu @rpautrat maybe we could add this as an extra to the tutorial? https://kornia-tutorials.readthedocs.io/en/latest/_nbs/line_detection_and_matching_sold2.html

or even make a separate tutorial with more introspection of the model.

**Comment by ogencoglu** [comment]
Thanks for the swift reply!

@edgarriba  yeah that would be great. Also it would be interesting to understand the relationship of model configs, line heatmap and line probabilities. For example, is the heatmap somehow "normalized" meaning that are heatmaps from different model configs comparable etc.?

**Comment by edgarriba** [comment]
as for @ogencoglu snippet i can say yes, seems that the output is in pixel space because later grid_sample takes in normalized flow grids

**Comment by rpautrat** [comment]
> @ogencoglu @rpautrat maybe we could add this as an extra to the tutorial? https://kornia-tutorials.readthedocs.io/en/latest/_nbs/line_detection_and_matching_sold2.html
> 
> or even make a separate tutorial with more introspection of the model.

Yes, I can do a PR to the tutorial to add the code snippet.

**Comment by rpautrat** [comment]
> Thanks for the swift reply!
> 
> @edgarriba yeah that would be great. Also it would be interesting to understand the relationship of model configs, line heatmap and line probabilities. For example, is the heatmap somehow "normalized" meaning that are heatmaps from different model configs comparable etc.?

In theory, the line heatmap is somewhat comparable and has the same scale between images. But it is quite approximate probably, and the resulting probability would be better suited for comparing lines within the same image (e.g. to select the top k lines).
